cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,151 @@
+
+Users rely on toFQDN policies to enforce network policies against traffic to remote destinations outside the cluster, typically to blob storage / other services on the internet. Rolling out cilium agent should not result in traffic being dropped. Introducing a high availablility (HA) mode will allow for adoption of toFQDN network policies in critical environments.
+
+## Goals
+
+* Introduce a streaming gRPC API for exchanging FQDN related policy information and endpoint related metadata.
+* Introduce standalone DNS proxy (SDP) that binds on the same port as built-in proxy with SO_REUSEPORT and uses the above mentioned API to notify agent of new DNS resolutions.
+
+## Non-Goals
+
+* Updating new DNS <> IP mappings when agent is down is out of scope
+
+## Proposal
+
+### Overview
+
+TODO : Add Image
+
+There are two parts to enforcing toFQDN network policy. L4 policy enforcement against IP adresses resolved from a FQDN and policy enforcement on DNS requests (L7 DNS policy). In order to enforce L4 policy, per endpoint policy bpf maps need to be updated. We'd like to avoid multiple processes writing entries to policy maps, so the standalone DNS proxy (SDP) needs a mechansim to notify agent of newly resolved FQDN <> IP address mappings. This CFP proposes exposing a new gRPC streaming API from cilium agent to do this. Since the connection is bi-directional, cilium agent can re-use the same connection to notify the SDP of L7 DNS policy changes. Additionally SDP also needs to translate IP address to endpoint ID and identity in order to enforce policy by reusing the logic from agent's DNS proxy.
+
+### RPC Methods
+
+Method : UpdateMappings
+
+Request : 
+```
+message FQDNMapping {
+    string FQDN = 1;
+    repeated bytes IPS = 2;
+    int32 TTL = 3;
+    bytes client_ip = 4;
+}
+```
+Response : 
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : UpdatesDNSRules
+
+_rpc UpdatesDNSRules(stream DNSPolicyRules) returns (Result){}_
+
+Request : 
+```
+message FQDNSelector {
+  string match_name = 1;
+  string match_pattern = 2;
+}
+
+message DNSPolicyRule {
+  uint32 port = 1;
+  string selector_string = 2;
+  repeated FQDNSelector port_rules= 3;
+  repeated string match_labels = 4;
+}
+
+message DNSPolicyRules {
+  uint64 endpoint_id = 1;
+  repeated DNSPolicyRule rules = 2;
+}
+```
+Response : 
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : GetEndpointMetadataByIP
+
+Request : 
+```
+message FQDNSelector {
+  string ip = 1;
+}
+```
+Response : 
+```
+message Result {
+    uint32 endpoint_id = 1;
+    uint32 identity = 2;
+}
+```
+
+### Tracking policy updates to SDP instances
+
+Since SDP can be deployed out of band and users can choose to completely disable built-in proxy to run multiple instances of SDP, agent should be prepared to handle multiple instances. In order to ensure all instances  have upto date policy revisions, agent will maintain a mapping of ACKed policy revision numbers against stream ID.
```

Comment:
> run multiple instances of SDP

How will multiple instances on the same node be managed? If we are assuming daemonsets, AFAIK you cannot scale them to multiple instances on the same node.

Reply:
multiple instances refers to CA dnsproxy and SDP. In another instance, while upgrading SDP, there can be 2 instances of SDP running at same time and still design allows that to be handled.